### PR TITLE
WIP: core, eth: initial exposure of the snap/1 protocol

### PR DESCRIPTION
**Pull request by karalabe**
WIP

### CODE REVIEW

**Comment by holiman** [comment]
WCGW :shipit:

**Comment by fjl** [comment]
I think it'd be better to put the snapshot protocol into its own package. It will be simpler to test this if it's standalone. To make this work, we probably need to create a shared package (maybe "eth/internal/ethpeer") for the peerset stuff.

**Comment by karalabe** [comment]
That's exactly what I'm afraid of. That I need to create a shared generic abstract package that can work with multiple protocols. I'm not capable of focusing on implementing the snapshot sync and reworking Geth's internals at the same time. Lemme first finish an operational snap protocol, then we can debate where to take it after we know what's needed. Until then it's just wasted time abstracting stuff out.

**Comment by holiman** [comment]
I seem to consistently fail to `ctrl-c`, always having to resort to a full blown panic in the end.

**Comment by karalabe** [comment]
I haven't focused on these things at all yet :) Shutdown logic will need to
be fully redone with the concurrency. The code is mostly for bechmarking
and testing.

File: core/state/statedb.go [hunk 1]
```diff
@@ -311,22 +311,38 @@ func (s *StateDB) GetState(addr common.Address, hash common.Hash) common.Hash {
...
 	if trie == nil {
 		return proof, errors.New("storage trie for requested address does not exist")
 	}
 	err := trie.Prove(crypto.Keccak256(key.Bytes()), 0, &proof)
-	return [][]byte(proof), err
+	return proof, err
+}
+
+// GetStorageProofByHash returns the Merkle proof for given storage slot.
+func (s *StateDB) GetStorageProofByHash(a common.Address, key common.Hash) ([][]byte, error) {
```
Comment (holiman) [inline core/state/statedb.go:338]: Are you trolling? I've :eye: :eye: but cannot spot any difference between `GetStorageProofByHash` and `GetStorageProof`?

File: eth/protocols/snap/sync.go [hunk 2]
```diff
@@ -0,0 +1,718 @@
...
+	response := &accountResponse{
+		hashes:   hashes,
+		accounts: accounts,
+		nodes:    db,
+		trie:     tr,
+		bounds:   bounds,
+	}
+	select {
+	//case <-req.cancel:
+	case s.accountResps <- response:
```
Comment (holiman) [inline eth/protocols/snap/sync.go:541]: hmm.. do we just drop this on the floor if the other side isn't listening? why would this happen? Shouldn't we at minimum log that we're dropping a batch of verified, useful data ? Or is it buffered?

File: eth/protocols/snap/sync.go [hunk 3]
```diff
@@ -0,0 +1,718 @@
...
+
+		for j < len(req.hashes) && !bytes.Equal(hash, req.hashes[j][:]) {
+			codes = append(codes, nil)
+			j++
+		}
+		if j < len(req.hashes) {
+			codes = append(codes, bytecodes[i])
+			j++
+			continue
+		}
```
Comment (holiman) [inline eth/protocols/snap/sync.go:662]: This looks very fishy.... 
So if the `0th` code mismatches. Then codes will be `codes=[nil], j=1`. Then you check `hash` against `req.hashes[1]`.  So for every requested hash, you're searching forward.. 

So it looks like you _almost_ allow codes to be returned in arbitrary order. But if that's the case, I think you're missing half the search-space?

But if you expect the hashes to be in the same order as the request, then it looks very odd that you don't update the `hash` when you move forward among the returned codes.

File: eth/protocols/snap/sync.go [hunk 4]
```diff
@@ -0,0 +1,718 @@
...
+	if !ok || req.id != id {
+		peer.Log().Warn("Unexpected account range packet")
+		s.lock.Unlock()
+		return nil
+	}
+	delete(s.accountReqs, peer.id)
+	root := s.root
+	s.lock.Unlock()
+
+	// If the response is unavailable snapshot, forward to the requester
```
Comment (holiman) [inline eth/protocols/snap/sync.go:500]: > If the response is unavailable snapshot

foo?

File: eth/protocols/snap/sync.go [hunk 5]
```diff
@@ -0,0 +1,718 @@
...
+	}
+	// Reconstruct a partial trie from the response and verify it
+	keys := make([][]byte, len(hashes))
+	for i, key := range hashes {
+		keys[i] = common.CopyBytes(key[:])
+	}
+	nodes := make(light.NodeList, len(proof))
+	for i, node := range proof {
+		nodes[i] = node
+	}
```
Comment (holiman) [inline eth/protocols/snap/sync.go:516]: Is all the copying needed?

File: eth/protocols/snap/sync.go [hunk 6]
```diff
@@ -0,0 +1,718 @@
...
+				// Sync was interrupted, restart next cycle at the current account,
+				// but leave next slot at wherever we were.
+				//
+				// TODO(karalabe): Special case account deletion in the next cycle or proof-lessness, musn't write
+				s.nextAcc = res.hashes[i]
+				break
+			}
+			// Account processed fully (may still be incomplete, but that's for
+			// trie node sync to complete), push the next account marker
+			s.nextAcc = common.BigToHash(new(big.Int).Add(res.hashes[i].Big(), big.NewInt(1)))
```
Comment (holiman) [inline eth/protocols/snap/sync.go:368]: This assumes that the accounts delivered are sorted. Is that guaranteed somewhere?

Comment (holiman) [reply eth/protocols/snap/sync.go:368]: I don't think it is. I added that as a testcase to the trie prover malicious-tests PR

File: core/state/statedb.go (hunk 1 above)
Comment (karalabe) [reply core/state/statedb.go:338]: The key in this one is a hash, in the other one it's hashed first. Every code we have uses addresses so we can't get rid of the other method, but the snapshot only knows about hashes, so we need this one too.

File: eth/protocols/snap/sync.go (hunk 2 above)
Comment (karalabe) [reply eth/protocols/snap/sync.go:541]: There's no `default` branch, so it blocks until it's read. The cancel case is also needed, but I need to implement some retry mechanism first, that's why I commented it out currently.

File: eth/protocols/snap/sync.go (hunk 4 above)
Comment (karalabe) [reply eth/protocols/snap/sync.go:500]: If the response is "this snapshot is not available at the remote peer", forward to the requested.

Though I agree it needs a few takes to get it, I'll rephrase

File: eth/protocols/snap/sync.go (hunk 5 above)
Comment (karalabe) [reply eth/protocols/snap/sync.go:516]: It's a slice conversion. But you're right, that `common.CopyBytes` is probably not needed, just the `key[:]` thing

File: eth/protocols/snap/sync.go (hunk 6 above)
Comment (karalabe) [reply eth/protocols/snap/sync.go:368]: It must be sorted, that's a protocol requirement. Not sure if it's enforced anywhere though. If not, we should.

File: eth/protocols/snap/sync.go [hunk 7]
```diff
@@ -0,0 +1,718 @@
...
+		case s.accountResps <- nil:
+		}
+		return nil
+	}
+	// Reconstruct a partial trie from the response and verify it
+	keys := make([][]byte, len(hashes))
+	for i, key := range hashes {
+		keys[i] = common.CopyBytes(key[:])
+	}
+	nodes := make(light.NodeList, len(proof))
```
Comment (holiman) [inline eth/protocols/snap/sync.go:513]: This is a bit inefficient. Perhaps it doesn't matter for accounts, but it does for storage. If the set is small enough to fit in a a batch, we shouldn't provide the proofs. For small storage tries, that could make quite a big difference. 
This diff fixes it for me

```diff
diff --git a/eth/protocols/snap/sync.go b/eth/protocols/snap/sync.go
index 298dc3b479..4818f63b10 100644
--- a/eth/protocols/snap/sync.go
+++ b/eth/protocols/snap/sync.go
@@ -510,13 +510,16 @@ func (s *Syncer) OnAccounts(peer PeerIF, id uint64, hashes []common.Hash, accoun
        for i, key := range hashes {
                keys[i] = common.CopyBytes(key[:])
        }
-       nodes := make(light.NodeList, len(proof))
-       for i, node := range proof {
-               nodes[i] = node
+       var proofDb ethdb.KeyValueReader
+       if len(proof) > 0{
+               nodes := make(light.NodeList, len(proof))
+               for i, node := range proof {
+                       nodes[i] = node
+               }
+               proofDb = nodes.NodeSet()
        }
-       proofdb := nodes.NodeSet()
 
-       db, tr, err := trie.VerifyRangeProof(root, req.origin[:], keys, accounts, proofdb, proofdb)
+       db, tr, err := trie.VerifyRangeProof(root, req.origin[:], keys, accounts, proofDb, proofDb)
        if err != nil {
                return err
        }
```

File: eth/protocols/snap/sync.go [hunk 8]
```diff
@@ -0,0 +1,718 @@
...
+
+	// If the request is stale, discard it
+	s.lock.Lock()
+	req, ok := s.bytecodeReqs[peer.id]
+	if !ok || req.id != id {
+		peer.Log().Warn("Unexpected bytecode packet")
+		s.lock.Unlock()
+		return nil
+	}
+	delete(s.storageReqs, peer.id)
```
Comment (holiman) [inline eth/protocols/snap/sync.go:640]:
```suggestion
	delete(s.bytecodeReqs, peer.id)
```

File: core/rawdb/accessors_snapshot.go [hunk 9]
```diff
@@ -118,3 +118,24 @@ func DeleteSnapshotJournal(db ethdb.KeyValueWriter) {
 		log.Crit("Failed to remove snapshot journal", "err", err)
 	}
 }
+
+// ReadSanpshotSyncStatus retrieves the serialized sync status saved at shutdown.
+func ReadSanpshotSyncStatus(db ethdb.KeyValueReader) []byte {
```
Comment (holiman) [inline core/rawdb/accessors_snapshot.go:123]: s/ReadSanpshotSyncStatus/ReadSnapshotSyncStatus

**Comment by holiman** [comment]
The 'server' should not blindly trust that `bytes` is `500K`, or some other 'sane' limit. If a client requests the entire account trie and sends `50G`, then I think the current implementation will just `OOM`. 

The snap protocol documentation should also be updated to highlight that the server should enforce a server-side limit to protect against this

**Comment by karalabe** [comment]
The request being served is capped to 1MB in handler.go.

**Comment by holiman** [comment]
Sorry, just saw `softResponseLimit` (`2Mb`)

File: eth/protocols/snap/sync.go [hunk 10]
```diff
@@ -0,0 +1,1690 @@
...
+		s.lock.Unlock()
+		return nil
+	}
+	root := s.root
+	s.lock.Unlock()
+
+	// Reconstruct a partial trie from the response and verify it
+	keys := make([][]byte, len(hashes))
+	for i, key := range hashes {
+		keys[i] = common.CopyBytes(key[:])
```
Comment (holiman) [inline eth/protocols/snap/sync.go:1378]: Do you really need to copy this? The `VerifyRangeProof` will internally do `TryUpdate`, which will copy+expand the keys with `k := keybytesToHex(key)`, leaving the original one intact, afaict

**Comment by holiman** [comment]
Similarly to the softresponselimit, we could have a lower floor. Right now, we'll behave a bit disorderly if the `bytes` is zero, e..g not deliver accounts (but supply proof for the first one), but for bytecode queries we will deliver one.

File: eth/protocols/snap/sync.go [hunk 11]
```diff
@@ -0,0 +1,1690 @@
...
+	s.peerJoin.Send(peer.id)
+	return nil
+}
+
+// Unregister injects a new data source into the syncer's peerset.
+func (s *Syncer) Unregister(id string) error {
+	// Remove all traces of the peer from the registry
+	s.lock.Lock()
+	if _, ok := s.peers[id]; !ok {
+		log.Error("Snap peer not registered", "id", id)
```
Comment (rjl493456442) [inline eth/protocols/snap/sync.go:345]: The lock is not released here

File: eth/protocols/snap/sync.go [hunk 12]
```diff
@@ -0,0 +1,1690 @@
...
+	// Task filling persisted, push it the chunk marker forward to the first
+	// account still missing data.
+	for i, hash := range res.hashes {
+		if task.needCode[i] || task.needState[i] {
+			return
+		}
+		task.Next = common.BigToHash(new(big.Int).Add(hash.Big(), big.NewInt(1)))
+	}
+	// All accounts marked as complete, track if the entire task is done
+	task.done = !res.cont
```
Comment (rjl493456442) [inline eth/protocols/snap/sync.go:1321]: I'm not sure it's correct. `cont` is the flag to the whole trie, not for the divided range.

Although we will set the `cont` to false if the response contains the element in the NEXT range, but in the protocol, the server won't return elements exceed the LIMIT the request specified.

File: eth/protocols/snap/sync.go (hunk 11 above)
Comment (karalabe) [reply eth/protocols/snap/sync.go:345]: Nice catch!

**Comment by karalabe** [comment]
Superseded by https://github.com/ethereum/go-ethereum/pull/21482

**Pull request closed without merging**